oauth-keycloak-authorizer/src/main/java/io/strimzi/kafka/oauth/server/authorizer/KeycloakRBACAuthorizer.java

```diff
@@ -0,0 +1,416 @@
+                } else {
+                    log.warn("Unexpected status while fetching authorization data - will retry next time: " + e.getMessage());
+                }
+            }
+            if (authz != null) {
+                // store authz grants in the token so they are available for subsequent requests
+                token.setPayload(authz);
+            }
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("authorize(): " + authz);
+        }
+
+        //
+        // Iterate authorization rules and try to find a match
+        //
+
+        Iterator<JsonNode> it = authz.iterator();
+        while (it.hasNext()) {
+            JsonNode permission = it.next();
+            String name = permission.get("rsname").asText();
+            ResourceSpec resourceSpec = ResourceSpec.of(name);
+            if (resourceSpec.match(clusterName, resource.resourceType().name(), resource.name())) {
+
+                ScopesSpec grantedScopes = ScopesSpec.of(
+                        validateScopes(
+                            JSONUtil.asListOfString(permission.get("scopes"))));
+
+                if (grantedScopes.isGranted(operation.name())) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Authorization GRANTED - cluster: " + clusterName + ",user: " + session.principal() + ", operation: " + operation +
+                                ", resource: " + resource + "\nGranted scopes for resource (" + resourceSpec + "): " + grantedScopes);
+                    }
+                    return true;
+                }
+            }
+        }
+
+        return delegateIfRequested(session, operation, resource, authz);
+    }
+
+    static List<ScopesSpec.AuthzScope> validateScopes(List<String> scopes) {
+        List<ScopesSpec.AuthzScope> enumScopes = new ArrayList<>(scopes.size());
+        for (String name: scopes) {
+            try {
+                enumScopes.add(ScopesSpec.AuthzScope.valueOf(name));
+            } catch (Exception e) {
+                log.warn("[IGNORED] Invalid scope detected in authorization scopes list: " + name);
+            }
+        }
+        return enumScopes;
+    }
+
+    boolean delegateIfRequested(RequestChannel.Session session, Operation operation, Resource resource, JsonNode authz) {
+        if (delegateToKafkaACL) {
+            boolean granted = super.authorize(session, operation, resource);
+
+            if (log.isDebugEnabled()) {
+                String status = granted ? "GRANTED" : "DENIED";
+                log.debug("Authorization " + status + " by ACL - user: " + session.principal() + ", operation: " + operation + ", resource: " + resource);
+            }
+            return granted;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Authorization DENIED - cluster: " + clusterName + ", user: " + session.principal() + ", operation: " + operation + ", resource: " + resource + "\n permissions: " + authz);
+        }
```

Comment:
I think we need to be a bit careful with the logging here.

For auditing purposes it's likely that users might be more interested when authorization is denied than when it's granted. Kafka attempted to solve that by logging deny at a higher level. Although that doesn't really work (because `authorize` can also be used for filtering what's included in protocol responses, as mentioned in https://cwiki.apache.org/confluence/display/KAFKA/KIP-504+-+Add+new+Java+Authorizer+Interface#KIP-504-AddnewJavaAuthorizerInterface-AclAuthorizer), I think we need to distinguish deny in a way which users can configure their logging for. i.e. either level, or logger name. But we should should also be aware when we've delegated result will also (probably) have been logged by the delegate.

Reply:
Good point. I gave no thought to possible auditing requirements here. Usually it has to be a formatted output that's easy to parse. For now I'll just make separate loggers and use debug level to log. Logging can then be configured separately for GRANT and DENY logging.

Reply:
There are now two additional logs - one for granted permissions and one for denied. One needs to enable debug logging on them to see logging output.